subsys/bluetooth/shell/audio.c

```diff
@@ -795,14 +795,22 @@ static void add_sink(struct bt_audio_ep *ep, uint8_t index)
 {
 	shell_print(ctx_shell, "Sink #%u: ep %p", index, ep);
 
+#if CONFIG_BT_AUDIO_UNICAST_CLIENT_ASE_SNK_COUNT == 0
+	__ASSERT(false, "No sinks configured");
+#else
 	snks[index] = ep;
+#endif
```

Comment:
Do you know if it is a compiler, or linker issue? If it is the latter, then perhaps a more simple
```c
	if (CONFIG_BT_AUDIO_UNICAST_CLIENT_ASE_SNK_COUNT > 0) {
		snks[index] = ep;
	}
```

Would be better?

Or even better, use `ARRAY_SIZE(snks)`. 

This is actually somewhat similar to a commit I recently did: https://github.com/zephyrproject-rtos/zephyr/pull/53291/commits/e59172243a51b6666c5c6b6d93b1798b370d5932

Perhaps the entire `snks` array (and the `srcs` array) should be guarded, and all functions related to it. There's no reason to keep any of it compiled if the respective array sizes (e.g. `CONFIG_BT_AUDIO_UNICAST_CLIENT_ASE_SNK_COUNT`) are 0. 

I think the current proposed solution needs to be properly implemented by either guarding everything (and don't use ASSERT), or keep the code as is (as the current code don't have any issues unless you change the Kconfig options).

Reply:
I'm using the 0.16.0 SDK compiler, which generates warnings for the bluetooth.audio_shell.no_client_ase_snk and bluetooth.audio_shell.no_client_ase_src tests (here's the compiler output for no_client_ase_snk):
```
/home/keithp/zephyrproject/zephyr/subsys/bluetooth/shell/audio.c: In function ‘cmd_config’:
/home/keithp/zephyrproject/zephyr/subsys/bluetooth/shell/audio.c:953:26: error: array subscript index is outside array bounds of ‘struct bt_audio_ep *[0]’ [-Werror=array-bounds]
  953 |                 ep = snks[index];
      |                      ~~~~^~~~~~~
/home/keithp/zephyrproject/zephyr/subsys/bluetooth/shell/audio.c:48:28: note: while referencing ‘snks’
   48 | static struct bt_audio_ep *snks[CONFIG_BT_AUDIO_UNICAST_CLIENT_ASE_SNK_COUNT];
      |                            ^~~~
In function ‘add_sink’,
    inlined from ‘discover_cb’ at /home/keithp/zephyrproject/zephyr/subsys/bluetooth/shell/audio.c:819:4,
    inlined from ‘discover_cb’ at /home/keithp/zephyrproject/zephyr/subsys/bluetooth/shell/audio.c:808:13:
/home/keithp/zephyrproject/zephyr/subsys/bluetooth/shell/audio.c:798:13: error: array subscript <unknown> is outside array bounds of ‘struct bt_audio_ep *[0]’ [-Werror=array-bounds]
  798 |         snks[index] = ep;
      |         ~~~~^~~~~~~
/home/keithp/zephyrproject/zephyr/subsys/bluetooth/shell/audio.c: In function ‘discover_cb’:
/home/keithp/zephyrproject/zephyr/subsys/bluetooth/shell/audio.c:48:28: note: while referencing ‘snks’
   48 | static struct bt_audio_ep *snks[CONFIG_BT_AUDIO_UNICAST_CLIENT_ASE_SNK_COUNT];
      |                            ^~~~
In function ‘add_sink’,
    inlined from ‘discover_all’ at /home/keithp/zephyrproject/zephyr/subsys/bluetooth/shell/audio.c:843:4,
    inlined from ‘discover_all’ at /home/keithp/zephyrproject/zephyr/subsys/bluetooth/shell/audio.c:832:13:
/home/keithp/zephyrproject/zephyr/subsys/bluetooth/shell/audio.c:798:13: error: array subscript <unknown> is outside array bounds of ‘struct bt_audio_ep *[0]’ [-Werror=array-bounds]
  798 |         snks[index] = ep;
      |         ~~~~^~~~~~~
/home/keithp/zephyrproject/zephyr/subsys/bluetooth/shell/audio.c: In function ‘discover_all’:
/home/keithp/zephyrproject/zephyr/subsys/bluetooth/shell/audio.c:48:28: note: while referencing ‘snks’
   48 | static struct bt_audio_ep *snks[CONFIG_BT_AUDIO_UNICAST_CLIENT_ASE_SNK_COUNT];
      |                            ^~~~
cc1: all warnings being treated as errors
```

And, yes, I agree that the correct solution is to elide all of this code when the configuration doesn't support these options. One reason is that zero-sized arrays are a GNU extension, not part of standard C. I'd be happiest if you'd take this as a bug report and fix it 'for real'; I'm just trying to work through toolchain adventures related to picolibc integration (not that this bug is associated with using picolibc, just that picolibc will appear with SDK version 0.16 when ready).

You can get a beta version of the 0.16 toolchain here: https://github.com/zephyrproject-rtos/sdk-ng/releases/tag/v0.16.0-beta1 -- I think that should exhibit this same issue, although I haven't tested that exactly toolchain version.